// ng2-scrollspy/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var core_1 = require('@angular/core');
var service_1 = require('./src/core/service');
var window_directive_1 = require('./src/core/window.directive');
var element_directive_1 = require('./src/core/element.directive');
var index_service_1 = require('./src/plugin/index.service');
__export(require('./src/core/service'));
__export(require('./src/core/window.directive'));
__export(require('./src/core/element.directive'));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    directives: [window_directive_1.ScrollSpyDirective, element_directive_1.ScrollSpyElementDirective],
    providers: [service_1.ScrollSpyService]
};
var ScrollSpyModule = (function () {
    function ScrollSpyModule(parentModule) {
        if (parentModule) {
            throw new Error('ScrollSpyModule already loaded; Import in root module only.');
        }
    }
    ScrollSpyModule.forRoot = function () {
        return {
            ngModule: ScrollSpyModule,
            providers: [service_1.ScrollSpyService, index_service_1.ScrollSpyIndexService]
        };
    };
    ScrollSpyModule = __decorate([
        core_1.NgModule({
            declarations: [window_directive_1.ScrollSpyDirective],
            exports: [window_directive_1.ScrollSpyDirective]
        }),
        __param(0, core_1.Optional()),
        __param(0, core_1.SkipSelf()), 
        __metadata('design:paramtypes', [ScrollSpyModule])
    ], ScrollSpyModule);
    return ScrollSpyModule;
}());
exports.ScrollSpyModule = ScrollSpyModule;
